File: ai/nikucooker_ai/protocol/codec.py
"""Newline-delimited JSON framing, mirroring pkg/protocol/codec.go.

One message is exactly one line of UTF-8 JSON. Go writes requests on the
worker's stdin; the worker writes events on a private handle to the real stdout
(see ``nikucooker_ai.__main__``), never through ``sys.stdout``.
"""

from __future__ import annotations

import json
from typing import Any, BinaryIO

from pydantic import BaseModel, ValidationError

from nikucooker_ai.protocol.errors import ProtocolError
from nikucooker_ai.protocol.models import (
    PROTOCOL_VERSION,
    ErrorEvent,
    FatalEvent,
    ProgressEvent,
    ReadyEvent,
    Request,
    ResultEvent,
)

#: Bounds a single message, matching the Go reader.
MAX_LINE_BYTES = 64 << 20  # 64 MiB

#: Event type -> model, used to dispatch an incoming event.
_EVENT_MODELS: dict[str, type[BaseModel]] = {
    "progress": ProgressEvent,
    "result": ResultEvent,
    "error": ErrorEvent,
    "ready": ReadyEvent,
    "fatal": FatalEvent,
}


class ProtocolDecodeError(Exception):
    """A message could not be decoded.

    Always a protocol violation rather than bad user input, so it is reported and
    never retried.
    """


class ProtocolLineTooLong(ProtocolDecodeError):
    """A message exceeded :data:`MAX_LINE_BYTES`.

    Kept separate from a generic decode error because the consequence differs: an
    oversized line desynchronises the stream, so it is fatal rather than
    skippable.
    """


class ProtocolReader:
    """Reads newline-delimited JSON messages from a binary stream."""

    def __init__(self, stream: BinaryIO, max_bytes: int = MAX_LINE_BYTES) -> None:
        self._stream = stream
        self._max = max_bytes

    def read_line(self) -> bytes | None:
        """Returns the next non-blank message, or ``None`` at end of stream.

        Blank lines are skipped rather than reported: a stray newline is a common
        artifact of a shell pipeline and carries no meaning. A trailing carriage
        return is stripped so a Windows writer behaves identically.
        """
        while True:
            raw = self._stream.readline(self._max + 1)
            if not raw:
                return None
            if len(raw) > self._max:
                raise ProtocolLineTooLong(
                    f"message exceeds {self._max} bytes; the stream cannot be resynchronised"
                )
            raw = raw.rstrip(b"\r\n")
            if not raw.strip():
                continue
            return raw

    def read_request(self) -> Request | None:
        """Reads and validates the next request."""
        line = self.read_line()
        if line is None:
            return None
        return decode_request(line)


class ProtocolWriter:
    """Writes newline-delimited JSON messages to a binary stream.

    Every write is flushed. A buffered message that has not reached the pipe is
    indistinguishable from a hang to the other side, and this protocol has no
    other liveness signal.
    """

    def __init__(self, stream: BinaryIO) -> None:
        self._stream = stream

    def write(self, message: BaseModel | dict[str, Any]) -> None:
        payload = message if isinstance(message, dict) else dump_wire(message)
        line = canonical_json(payload)
        self._stream.write(line + b"\n")
        self._stream.flush()

    def write_event(self, event: BaseModel) -> None:
        """Writes an event, refusing anything that is not one.

        The guard exists because the failure it prevents is silent: a request
        written to the event stream would be read by the core as an unknown
        message type, and the symptom would appear far from the cause.
        """
        etype = getattr(event, "type", None)
        if etype not in _EVENT_MODELS:
            raise ValueError(f"not an event: {type(event).__name__}")
        self.write(event)

    def write_error(self, request_id: str, error: ProtocolError) -> None:
        self.write(ErrorEvent(id=request_id, error=error))

    def write_fatal(self, error: ProtocolError) -> None:
        self.write(FatalEvent(error=error))


def decode_request(line: bytes) -> Request:
    """Decodes and validates a request line."""
    try:
        payload = json.loads(line)
    except json.JSONDecodeError as exc:
        raise ProtocolDecodeError(f"malformed request: {exc}") from exc

    if not isinstance(payload, dict):
        raise ProtocolDecodeError("request is not a JSON object")

    version = payload.get("v")
    if version != PROTOCOL_VERSION:
        raise ProtocolDecodeError(
            f"unsupported protocol version: got {version!r}, want {PROTOCOL_VERSION}"
        )

    try:
        return Request.model_validate(payload)
    except ValidationError as exc:
        raise ProtocolDecodeError(f"malformed request: {exc}") from exc


def decode_event(line: bytes) -> BaseModel:
    """Decodes an event line into its concrete model."""
    try:
        payload = json.loads(line)
    except json.JSONDecodeError as exc:
        raise ProtocolDecodeError(f"malformed event: {exc}") from exc

    if not isinstance(payload, dict):
        raise ProtocolDecodeError("event is not a JSON object")

    version = payload.get("v")
    if version != PROTOCOL_VERSION:
        raise ProtocolDecodeError(
            f"unsupported protocol version: got {version!r}, want {PROTOCOL_VERSION}"
        )

    etype = payload.get("type")
    model = _EVENT_MODELS.get(etype) if isinstance(etype, str) else None
    if model is None:
        raise ProtocolDecodeError(f"unknown event type: {etype!r}")

    try:
        return model.model_validate(payload)
    except ValidationError as exc:
        raise ProtocolDecodeError(f"malformed {etype} event: {exc}") from exc


def dump_wire(model: BaseModel) -> dict[str, Any]:
    """Serialises a model into its wire form.

    ``exclude_none`` mirrors Go's ``omitempty`` on pointers and optionals. It does
    not mirror ``omitempty`` on a plain Go string, which also drops ``""``: this
    side would emit ``"message": ""`` where Go emits nothing. The difference is
    benign — Go decodes both to the empty string — and closing it would mean
    dropping any empty string field, including required ones that Go does emit.
    """
    return model.model_dump(mode="json", exclude_none=True)


def canonical_json(value: Any) -> bytes:
    """Renders a value in canonical form.

    Object keys sorted, no insignificant whitespace, non-ASCII preserved.

    ``ensure_ascii=False`` is required for agreement with Go: the default escapes
    every non-ASCII character, so Japanese subtitle text would hash differently on
    the two sides. Go's equivalent switch is ``SetEscapeHTML(false)`` — see
    pkg/protocol/canonical.go.
    """
    return json.dumps(
        value,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    ).encode("utf-8")

File: pkg/protocol/codec.go
package protocol

import (
	"bufio"
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sync"
)

// MaxLineBytes bounds a single protocol line.
//
// This limit is not arbitrary and the reader is not a bufio.Scanner. Scanner's
// default token limit is 64 KiB and it stops silently at the first line that
// exceeds it — and an ASR result for a two-hour video exceeds that by a factor
// of thirty. The failure is a truncated stream with no error, which is why the
// reader below accumulates explicitly and refuses anything oversized.
const MaxLineBytes = 64 << 20 // 64 MiB

// ErrLineTooLong is returned for a line exceeding MaxLineBytes.
//
// Exceeding the limit is a protocol violation, not a recoverable condition: the
// stream is desynchronised at that point, so continuing would mean misparsing
// whatever follows.
var ErrLineTooLong = errors.New("protocol: message exceeds maximum line length")

// Reader reads newline-delimited JSON messages.
//
// It is not safe for concurrent use; one reader belongs to one goroutine.
type Reader struct {
	r   *bufio.Reader
	max int
}

// NewReader wraps r for protocol reading.
func NewReader(r io.Reader) *Reader {
	return &Reader{r: bufio.NewReaderSize(r, 64<<10), max: MaxLineBytes}
}

// ReadLine returns the next non-empty message, with the line terminator
// stripped.
//
// Blank lines are skipped rather than reported: a stray newline is a common
// artifact of a shell pipeline and carries no meaning. A trailing carriage
// return is stripped so that a worker running on Windows behaves identically to
// one on Unix.
func (r *Reader) ReadLine() ([]byte, error) {
	for {
		line, err := r.readBounded()
		if err != nil {
			return nil, err
		}
		line = bytes.TrimRight(line, "\r\n")
		if len(bytes.TrimSpace(line)) == 0 {
			continue
		}
		return line, nil
	}
}

// readBounded accumulates one line, refusing to grow past the limit.
func (r *Reader) readBounded() ([]byte, error) {
	var buf []byte
	for {
		chunk, err := r.r.ReadSlice('\n')
		// append copies immediately: chunk aliases the reader's internal buffer
		// and is invalidated by the next read.
		buf = append(buf, chunk...)

		if len(buf) > r.max {
			return nil, fmt.Errorf("%w: limit %d bytes", ErrLineTooLong, r.max)
		}
		if err == nil {
			return buf, nil
		}
		if errors.Is(err, bufio.ErrBufferFull) {
			continue
		}
		if errors.Is(err, io.EOF) && len(buf) > 0 {
			// A final line without a terminator is still a message.
			return buf, nil
		}
		return nil, err
	}
}

// Writer writes newline-delimited JSON messages.
//
// Safe for concurrent use: the core may emit progress from a reader goroutine
// while a request writer is separate, and interleaving two half-lines would
// desynchronise the stream irrecoverably.
type Writer struct {
	mu sync.Mutex
	w  *bufio.Writer
}

// NewWriter wraps w for protocol writing.
func NewWriter(w io.Writer) *Writer {
	return &Writer{w: bufio.NewWriterSize(w, 64<<10)}
}

// WriteMessage marshals v onto a single line and flushes it.
//
// The flush is unconditional. A buffered message that has not reached the pipe
// is indistinguishable from a hang to the other side, and this protocol has no
// other liveness signal.
func (w *Writer) WriteMessage(v any) error {
	payload, err := json.Marshal(v)
	if err != nil {
		return fmt.Errorf("protocol: marshal: %w", err)
	}
	// encoding/json escapes newlines inside strings, so this can only trip if a
	// type implements MarshalJSON and emits one. That would corrupt framing for
	// every subsequent message, so it is worth the check.
	if bytes.ContainsRune(payload, '\n') {
		return errors.New("protocol: marshalled message contains a newline")
	}

	w.mu.Lock()
	defer w.mu.Unlock()

	if _, err := w.w.Write(payload); err != nil {
		return fmt.Errorf("protocol: write: %w", err)
	}
	if err := w.w.WriteByte('\n'); err != nil {
		return fmt.Errorf("protocol: write: %w", err)
	}
	return w.w.Flush()
}

// DecodeRequest decodes a Go → worker request line.
func DecodeRequest(line []byte) (*Request, error) {
	var req Request
	if err := json.Unmarshal(line, &req); err != nil {
		return nil, fmt.Errorf("protocol: malformed request: %w", err)
	}
	if req.V != Version {
		return nil, fmt.Errorf("%w: got %d, want %d", ErrVersionMismatch, req.V, Version)
	}
	if req.ID == "" {
		return nil, errors.New("protocol: request has no id")
	}
	if req.Method == "" {
		return nil, errors.New("protocol: request has no method")
	}
	return &req, nil
}

// DecodeEvent decodes a worker → Go event line into its concrete type.
//
// The type tag is read first and the line is then decoded into the matching
// struct, so a field belonging to a different event type is a decode error
// rather than a silently ignored key.
func DecodeEvent(line []byte) (Event, error) {
	var probe struct {
		V    int    `json:"v"`
		Type string `json:"type"`
	}
	if err := json.Unmarshal(line, &probe); err != nil {
		return nil, fmt.Errorf("protocol: malformed event: %w", err)
	}
	if probe.V != Version {
		return nil, fmt.Errorf("%w: got %d, want %d", ErrVersionMismatch, probe.V, Version)
	}

	var (
		event Event
		err   error
	)
	switch probe.Type {
	case TypeProgress:
		var e ProgressEvent
		err, event = json.Unmarshal(line, &e), &e
	case TypeResult:
		var e ResultEvent
		err, event = json.Unmarshal(line, &e), &e
	case TypeError:
		var e ErrorEvent
		err, event = json.Unmarshal(line, &e), &e
	case TypeReady:
		var e ReadyEvent
		err, event = json.Unmarshal(line, &e), &e
	case TypeFatal:
		var e FatalEvent
		err, event = json.Unmarshal(line, &e), &e
	default:
		return nil, fmt.Errorf("protocol: unknown event type %q", probe.Type)
	}

	if err != nil {
		return nil, fmt.Errorf("protocol: malformed %s event: %w", probe.Type, err)
	}
	return event, nil
}

// ErrVersionMismatch reports a message whose protocol version this build does
// not speak.
var ErrVersionMismatch = errors.New("protocol: unsupported version")
